esp-alloc/src/lib.rs

```diff
@@ -61,35 +67,100 @@ use core::{
 };
 
 use critical_section::Mutex;
+use enumset::{EnumSet, EnumSetType};
 use linked_list_allocator::Heap;
 
+/// The global allocator instance
+#[global_allocator]
+pub static INSTANCE: EspHeap = EspHeap::empty();
+
+const NON_REGION: Option<HeapRegion> = None;
+
+#[derive(EnumSetType)]
+/// Describes the properties of a memory region
+pub enum MemoryCapability {
+    /// Memory must be internal; specifically it should not disappear when
+    /// flash/spiram cache is switched off
+    Internal,
+    /// Memory must be in SPI RAM
+    External,
+}
+
+/// A memory region to be used as heap memory
+pub struct HeapRegion {
+    heap: Heap,
+    capabilities: EnumSet<MemoryCapability>,
+}
+
+impl HeapRegion {
+    /// Create a new [HeapRegion] with the given capabilities
+    ///
+    /// # Safety
+    ///
+    /// - The supplied memory region must be available for the entire program
+    ///   (`'static`).
+    /// - The supplied memory region must be exclusively available to the heap
+    ///   only, no aliasing.
+    /// - `size > 0`.
+    pub unsafe fn new(
+        heap_bottom: *mut u8,
+        size: usize,
+        capabilities: EnumSet<MemoryCapability>,
+    ) -> Self {
+        let mut heap = Heap::empty();
+        heap.init(heap_bottom, size);
+
+        Self { heap, capabilities }
+    }
+}
+
+/// For esp-wifi
+#[doc(hidden)]
+#[no_mangle]
+pub extern "C" fn free_internal_heap() -> usize {
+    INSTANCE.free_caps(MemoryCapability::Internal.into())
+}
+
+/// For esp-wifi
+#[doc(hidden)]
+#[no_mangle]
+pub extern "C" fn allocate_from_internal_ram(size: usize) -> *mut u8 {
+    unsafe {
+        INSTANCE.alloc_caps(
+            MemoryCapability::Internal.into(),
+            Layout::from_size_align_unchecked(size, 4),
+        )
+    }
+}
+
+/// A memory allocator
+///
+/// In addition to what Rust's memory allocator can do it allows to allocate
+/// memory in regions satisfying specific needs.
 pub struct EspHeap {
-    heap: Mutex<RefCell<Heap>>,
+    heap: Mutex<RefCell<[Option<HeapRegion>; 3]>>,
```

Comment:
Any reason not to make this a const generic?

Reply:
Unfortunately, yes because the static can't be generic. Would be possible if we don't define the global allocator here and let the user define it but then `free_internal_heap` and `allocate_from_internal_ram` would be awkward to implement here

My thought was that 3 should be enough (max 1 x in DRAM, 1 x in DRAM2, 1 x PSRAM). On the other hand if the user just needs one, we don't waste too much (`EspHeap` is 100 bytes now)

Reply:
I meant making EspHeap generic but hard coding 3 in the static. (Unless that's what you meant then nevermind)

Reply:
> I meant making EspHeap generic but hard coding 3 in the static. (Unless that's what you meant then nevermind)

Ah yes - that should work

Reply:
On second thought, just ignore this comment. Given users can't use their own global allocator when using `esp-alloc` there's no point in adding the generic parameter. Users won't be able to make use of it.